File: cometchat_calls_sdk/android/src/main/kotlin/com/cometchatcalls/cometchatcalls_plugin/helper/GetMappedObject.kt
package com.cometchatcalls.cometchatcalls_plugin.helper

import com.cometchat.calls.model.*

/**
 * Created by Rohit Giri on 22/03/23.
 */
object GetMappedObject {
    fun rtcUserListMap(
        rtcUserList: ArrayList<RTCUser>?,
        methodName: String = ""
    ): HashMap<String, Any?>? {
        if (rtcUserList == null) return null
        if (rtcUserList.isEmpty()) return null
        return hashMapOf(
            "methodName" to methodName,
            "userList" to rtcUserList.map { e ->  rtcUserMap(e) }
        )
    }

    fun rtcUserMap(
        rtcUser: RTCUser?,
        methodName: String = ""
    ): HashMap<String, Any?>? {
        if (rtcUser == null) return null
        return hashMapOf(
            "uid" to rtcUser.uid,
            "name" to rtcUser.name,
            "avatar" to rtcUser.avatar,
            "jwt" to rtcUser.jwt,
            "resource" to rtcUser.resource,
            "methodName" to methodName
        )
    }

    fun audioModeListMap(
        audioModeList: ArrayList<AudioMode>?,
        methodName: String = ""
    ): HashMap<String, Any?>? {
        if (audioModeList == null) return null
        if (audioModeList.isEmpty()) return null
        return hashMapOf(
            "methodName" to methodName,
            "audioModeList" to audioModeList.map { e ->  audioModeMap(e) }
        )
    }

    fun audioModeMap(
        audioMode: AudioMode?,
    ): HashMap<String, Any?>? {
        if (audioMode == null) return null
        return hashMapOf(
            "mode" to audioMode.mode,
            "isSelected" to audioMode.isSelected
        )
    }

    fun callSwitchRequestInfoMap(
        callSwitchRequestInfo: CallSwitchRequestInfo?,
        methodName: String = ""
    ): HashMap<String, Any?>? {
        if (callSwitchRequestInfo == null) return null
        return hashMapOf(
            "sessionId" to callSwitchRequestInfo.sessionId,
            "requestInitiatedBy" to rtcUserMap(callSwitchRequestInfo.requestInitiatedBy),
            "requestAcceptedBy" to rtcUserMap(callSwitchRequestInfo.requestAcceptedBy),
            "methodName" to methodName
        )
    }

    fun rtcMutedUserMap(
        rtcMutedUser: RTCMutedUser?,
        methodName: String = ""
    ): HashMap<String, Any?>? {
        if (rtcMutedUser == null) return null
        return hashMapOf(
            "muted" to rtcUserMap(rtcMutedUser.muted),
            "mutedBy" to rtcUserMap(rtcMutedUser.mutedBy),
            "methodName" to methodName
        )
    }

    fun rtcRecordingInfoMap(
        rtcRecordingInfo: RTCRecordingInfo?,
        methodName: String = ""
    ): HashMap<String, Any?>? {
        if (rtcRecordingInfo == null) return null
        return hashMapOf(
            "recordingStarted" to rtcRecordingInfo.recordingStarted,
            "user" to rtcUserMap(rtcRecordingInfo.user),
            "methodName" to methodName
        )
    }

    fun cometchatExceptionMap(
        cometchatException: com.cometchat.calls.exceptions.CometChatException?,
        methodName: String = ""
    ): HashMap<String, Any?>? {
        if (cometchatException == null) return null
        return hashMapOf(
            "code" to cometchatException.code,
            "message" to cometchatException.message,
            "localizedMessage" to cometchatException.localizedMessage,
            "details" to cometchatException.details,
            "methodName" to methodName
        )
    }

    fun booleanMap(
        flag: Boolean,
        methodName: String = ""
    ): HashMap<String, Any?> {
        return hashMapOf(
            "code" to flag,
            "methodName" to methodName
        )
    }

    fun stringToMap(
        message: String = "",
        methodName: String = ""
    ): HashMap<String, Any?> {
        return hashMapOf(
            "message" to message,
            "methodName" to methodName
        )
    }

    fun successMap(
        msg: String,
        methodName: String
    ): HashMap<String, Any> {
        return hashMapOf(
            "message" to msg,
            "methodName" to methodName
        )
    }

}

File: cometchat_calls_sdk/ios/Classes/Helper/GetMappedObject.swift
import Foundation
import CometChatCallsSDK

class GetMappedObject {

    static func boolMap(
        flag: Bool,
        methodName: String
    ) -> [String: Any?] {
        return [
            "flag": flag,
            "methodName": methodName
        ]
    }
    
    static func rtcUserMap(
        rtcUser: NSDictionary,
        methodName: String = ""
    ) -> [String: Any?]{
        return [
            "uid": rtcUser["uid"],
            "name": rtcUser["name"],
            "avatar": rtcUser["avatar"],
            "jwt": rtcUser["jwt"],
            "resource": rtcUser["resource"],
            "methodName": methodName
        ]
    }
    
    static func rtcUserListMap(
        rtcUserList: NSArray,
        methodName: String = ""
    ) -> [String: Any?] {
        
        var userList: [Any?] = []
        rtcUserList.forEach { e in
            userList.append(rtcUserMap(rtcUser: e as! NSDictionary))
        }
        
        print(["userList": userList])
        
        return [
            "methodName": methodName,
            "userList": userList
        ]
    }
    
    static func audioModeMap(
        audioMode: NSDictionary?,
        methodName: String = ""
    ) -> [String: Any?]? {
        if audioMode == nil {
            print("iOS GetMappedObject audioModeMap: audioMode found null")
            return nil
        }
        return [
            "mode": audioMode!["mode"],
            "isSelected": audioMode!["isSelected"]
        ]
    }
    
    static func audioModeListMap(
        audioModeList: NSArray?,
        methodName: String = ""
    ) -> [String: Any?]? {
        
        if audioModeList == nil {
            print("iOS GetMappedObject audioModeListMap: audioModeList found null")
            return nil
        }
        
        var audioModeArray: [Any] = []
        audioModeList?.forEach({ e in
            audioModeArray.append(audioModeMap(audioMode: e as? NSDictionary))
        })
        
        return [
            "methodName": methodName,
            "audioModeList": audioModeArray
        ]
        
    }
    
    static func callSwitchRequestInfoMap(
        callSwitchRequestInfo: NSDictionary?,
        methodName: String
    ) -> [String: Any?]? {
        if callSwitchRequestInfo == nil {
            print("iOS GetMappedObject callSwitchRequestInfoMap: callSwitchRequestInfo found null")
            return nil
        }
        return [
            "sessionId": callSwitchRequestInfo!["sessionId"],
            "requestInitiatedBy": callSwitchRequestInfo!["requestInitiatedBy"],
            "requestAcceptedBy": callSwitchRequestInfo!["requestAcceptedBy"],
            "methodName": callSwitchRequestInfo!["methodName"]
        ]
    }
    
    static func rtcRecordingInfoMap (
        rtcRecordingInfo: NSDictionary?,
        methodName: String
    ) -> [String: Any?]? {
        
        guard let rtcRecordingInfo = rtcRecordingInfo else {
            print("iOS GetMappedObject rtcRecordingInfoMap: rtcRecordingInfo found null")
            return nil
        }
        
        return [
            "recordingStarted": rtcRecordingInfo["rtcRecordingInfo"],
            "user": rtcUserMap(rtcUser: rtcRecordingInfo["user"] as! NSDictionary),
            "methodName": methodName
        ]
    }
    
    static func rtcMutedUserMap (
        rtcMutedUser: NSDictionary?,
        methodName: String
    ) -> [String: Any?]? {
        
        guard let rtcMutedUser = rtcMutedUser else {
            print("iOS GetMappedObject rtcMutedUserMap: rtcMutedUser found null")
            return nil
        }
        
        return [
            "muted": rtcUserMap(rtcUser: rtcMutedUser["muted"] as! NSDictionary),
            "mutedBy": rtcUserMap(rtcUser: rtcMutedUser["mutedBy"] as! NSDictionary),
            "methodName": methodName
        ]
    }
    
    static func cometchatExceptionMap (
        code: String,
        message: String,
        localizedMessage: String,
        details: String,
        methodName: String
    ) -> [String: Any?] {
        return [
            "code": code,
            "message": message,
            "localizedMessage": localizedMessage,
            "details": details,
            "methodName": methodName
        ]
    }
    
    static func resultMap (
        msg: String?,
        methodName: String?
    ) -> [String: Any?] {
        return [
            "message": msg,
            "methodName": methodName
        ]
    }
    
}
